Fix for issue #452

mberning opened the pull request:
This change fixes #452 by URL encoding resource identifiers.

Captain-P-Goldfish commented on scim-sdk-common/src/main/java/de/captaingoldfish/scim/sdk/common/utils/EncodingUtils.java:
```diff
@@ -0,0 +1,32 @@
+   * @param value The value to be URL encoded.
+   * @return A URL encoded version of the input value
+   */
+  public static String urlEncode(String value)
```
could you move this into the class `UtilityMethods`?
I would prefer to have fewer than more classes of this kind :-)

mberning replied:
Sure, no problem, I will make the change.

Captain-P-Goldfish reviewed (commented):
I also found another problem:
The test-method in `ResourceEndpointTest#testGetResource()`. Try to modify the testcase like this:

```
final String userName = "chuck_norris";
final String id = EncodingUtils.urlEncode("$chuck- norris!");
```

You will see that a problem will arise that might cause trouble under certain circumstances. This should be fixed next.

mberning commented:
1. Fixed the areas where meta location provided by the developer implementation was being overridden and added unit tests
2. I looked at moving the encoding utility methods to the `UtilityMethods` class, but it is not currently available in `scim-sdk-server`. Not sure how you would like to proceed.
3. Fixing the bug where location was not being respected caused `ResourceTypeEndpointTest` to start failing. I am a bit concerned that it was depending on the old behavior. I started looking in to how the meta location is being set for the resource types, but wanted to get your input before proceeding.

Captain-P-Goldfish commented:
Seems like I broke CircleCi integration. I tried to repair it now:

1. There are still formatting causing the maven-build to error
2. 5 test error but I would be willing to fix them myself. You will have to modify  the schema files in order to fix these tests
3. The methods from the class `EncodingUtils` were not moved as discussed.

mberning commented:
> Seems like I broke CircleCi integration. I tried to repair it now:
> 
>     1. There are still formatting causing the maven-build to error. You can fix the formatting with `mvn formatter:format`
> 
>     2. 5 test errors but I would be willing to fix them myself. You will have to modify  the schema files in order to fix these tests
> 
>     3. The methods from the class `EncodingUtils` were not moved as discussed.

I pushed commits to resolve point 1 and 2. On point 3 I looked at moving the encoding utility methods to the `UtilityMethods` class, but it is not currently available in `scim-sdk-server`. Not sure how you would like to proceed.

Captain-P-Goldfish commented:
ignore point 3 :-)
I was somehow sure it was in the common module. my mistake

Captain-P-Goldfish commented:
Could you simply delete the location attributes in the schemas? Otherwise these schemas would be returned with erroneous locations in their location attribute from the /ResourceTypes endpoint.
Afterwards try to build with maven. You will notice that the build fails because I have entered checksums in the scim-sdk-common/pom.xml file to notice unintentional changes on these files.

mberning commented:
> Could you simply delete the location attributes in the schemas? Otherwise these schemas would be returned with erroneous locations in their location attribute from the /ResourceTypes endpoint. Afterwards try to build with maven. You will notice that the build fails because I have entered checksums in the scim-sdk-common/pom.xml file to notice unintentional changes on these files.

I gave this a try and ran into an issue with https://github.com/Captain-P-Goldfish/SCIM-SDK/blob/master/scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/schemas/ResourceType.java#L117

Seems that is is generating a location when none is provided, but is not aware of the current base URI.

mberning commented:
I pushed a fix and updated the checksums
